### PR TITLE
Cryoscope Measurement and Analysis

### PR DESCRIPTION
**Pull request by chellings**
This PR adds the measurement and analysis classes for cryoscope measurements. It collects all commits related to cryoscope that I could find (except the experimental feature BufferedCryoscopePulse, which we probably do not want to merge into the master yet).

The PR includes changes in pulse, pulse_library, and QuDev_transmon to introduce the pulse truncation feature, and it does some required modifications to tda and to the multitasking framework. In addition, the PR introduces a clock() method in BaseAnalysis to determine the clock frequency of a channel or AWG.

Inviting @stephlazar to review.
FYI @antsr @nathlacroix

### CODE REVIEW

File: pycqed/measurement/waveform_control/pulse.py [hunk 1]
```diff
@@ -173,6 +208,15 @@ def hashables(self, tstart, channel):
         raise NotImplementedError('hashables() not implemented for {}'
                                   .format(str(type(self))[1:-1]))
 
+    def common_hashables(self, tstart, channel):
```
Comment (stephlazar) [inline pycqed/measurement/waveform_control/pulse.py:211]: I don't know what this function does so I can't really review this part. But I think this has been used already on the measurement setups so it should be fine.

File: pycqed/measurement/waveform_control/pulse_library.py [hunk 2]
```diff
@@ -104,11 +105,9 @@ def chan_wf(self, channel, tvals):
             return np.zeros_like(tvals)
 
     def hashables(self, tstart, channel):
-        if channel not in self.channels:
-            return []
-        if self.pulse_off:
-            return ['Offpulse', self.algorithm_time() - tstart, self.length]
-        hashlist = [type(self), self.algorithm_time() - tstart]
+        hashlist = self.common_hashables(tstart, channel)
```
Comment (stephlazar) [inline pycqed/measurement/waveform_control/pulse_library.py:108]: Is this related to Cryoscope?

File: pycqed/measurement/waveform_control/pulse_library.py [hunk 3]
```diff
@@ -204,12 +203,12 @@ def chan_wf(self, channel, tvals):
     def hashables(self, tstart, channel):
         if channel in [self.I_channel, self.Q_channel]:
             return super().hashables(tstart, channel)
-        if self.pulse_off:
-            return ['Offpulse', self.algorithm_time() - tstart, self.length]
+        hashlist = self.common_hashables(tstart, channel)
```
Comment (stephlazar) [inline pycqed/measurement/waveform_control/pulse_library.py:206]: Is this related to Cryoscope?
Same question for all the changes in this file.

**Review by stephlazar — COMMENTED** [review]
My usual question: did you make sure you can still run Cryoscope, Cphase and Dynphase analyses with this branch?

File: pycqed/measurement/waveform_control/pulse_library.py (hunk 2 above)
Comment (chellings) [reply pycqed/measurement/waveform_control/pulse_library.py:108]: It adds the truncation parameters to the hashables, see definition of common_hashables. This is related to introducing the truncation feature in pulse, which is related to cryoscope. Since we now have the function common_hashables, other parameters could be moved into that function as well instead of replicating the code in every pulse.

File: pycqed/measurement/waveform_control/pulse_library.py (hunk 3 above)
Comment (chellings) [reply pycqed/measurement/waveform_control/pulse_library.py:206]: see my other response

File: pycqed/measurement/waveform_control/pulse.py (hunk 1 above)
Comment (chellings) [reply pycqed/measurement/waveform_control/pulse.py:211]: The parameters of the truncation feature in pulse must be part of the hashables of every pulse. This is required to ensure that reuse_waveforms and the sequence caching in pulse work correctly when the truncation feature is used.

**Comment by chellings** [comment]
> My usual question: did you make sure you can still run Cryoscope, Cphase and Dynphase analyses with this branch?

Yes, I did not have any issues.

**Review by stephlazar — APPROVED** [review]
